### PR TITLE
Add functional dependency to IsMember for better type inference

### LINKED ISSUE #26: Map: lookp return value type inference and multiple identical keys
In Map, if I write

```
myMap :: Map '[ "w" ':-> Int]
myMap = Ext (Var :: (Var "w")) 2 Empty

elemAtW = lookp (Var :: (Var "w")) myMap
```

GHC cannot deduce the type returned by lookp[1]. For my use case this makes Map not usable.  My first reasoning was that this was strange as, as I thought that in a map there can only be one key of a certain value (in this case type), and for that map the associated value is Int.

But then I tried creating a map with multiple identical keys, and it works:

```
myMap2 :: Map '[ "w" ':-> Int, "w" ':-> Int]
myMap2 = Ext (Var :: (Var "w")) (2::Int) $ Ext (Var :: (Var "w")) (4::Int) Empty
```

I see that `asMap` would transform this back a proper map, summing 2 and 4. Still, in terms of type inference this means that GHC cannot ever deduce the type returned by lookp ? Would in theory be possible to have a different implementation of Map where duplicate keys were not allowed and GHC could infer the type returned by lookp ?

[1]
```
• Overlapping instances for IsMember "w" t0 '[ "w" ':-> Int]
    arising from a use of ‘lookp’
  Matching instances:
    two instances involving out-of-scope types
    (use -fprint-potential-instances to see them all)
  (The choice depends on the instantiation of ‘t0’
   To pick the first instance above, use IncoherentInstances
   when compiling the other instance declarations)
• In the expression: lookp (Var :: (Var "w")) myMap
  In an equation for ‘elemAtW’:
      elemAtW = lookp (Var :: (Var "w")) myMap
```

### PR DESCRIPTION
**Pull request by jmorag**
Resolves #26

### CODE REVIEW

**Comment by jmorag** [comment]
Added another commit for better inference with `update`. Previously,
```haskell
ghci> m = Ext (Var @"z") 'b' $ Ext (Var @"w") 'a' Empty
ghci> m
{z :-> 'b', w :-> 'a'}
ghci> update m (Var @"w") 3
    • Could not deduce (Updatable
                          "w" t0 '[ "z" ':-> Char, "w" ':-> Char] n)
      from the context: (Updatable
                           "w" t '[ "z" ':-> Char, "w" ':-> Char] n,
                         Num t)
        bound by the inferred type for ‘it’:
                   forall t (n :: [Mapping Symbol *]).
                   (Updatable "w" t '[ "z" ':-> Char, "w" ':-> Char] n, Num t) =>
                   Map n
        at <interactive>:973:1-26
      The type variable ‘t0’ is ambiguous
    • In the ambiguity check for the inferred type for ‘it’
      To defer the ambiguity check to use sites, enable AllowAmbiguousTypes
      When checking the inferred type
        it :: forall t (n :: [Mapping Symbol *]).
              (Updatable "w" t '[ "z" ':-> Char, "w" ':-> Char] n, Num t) =>
              Map n
ghci> update m (Var ::Var "w") (3::Int)
update m (Var ::Var "w") (3::Int)
  :: Updatable "w" Int '[ "z" ':-> Char, "w" ':-> Char] n => Map n
```
With 686f64d,
```haskell
ghci> m = Ext (Var @"z") 'b' $ Ext (Var @"w") 'a' Empty
ghci> m
{z :-> 'b', w :-> 'a'}
ghci> update m (Var @"w") 3
{z :-> 'b', w :-> 3}
ghci> :t update m (Var @"w") 3
update m (Var @"w") 3 :: Num t => Map '[ "z" ':-> Char, "w" ':-> t]
```

File: src/Data/Type/Map.hs [hunk 1]
```diff
@@ -2,17 +2,17 @@
 The implementation is similar to that shown in the paper.
  "Embedding effect systems in Haskell" Orchard, Petricek 2014  -}
 
-{-# LANGUAGE TypeOperators, PolyKinds, DataKinds, KindSignatures,
-             TypeFamilies, UndecidableInstances, MultiParamTypeClasses,
+{-# LANGUAGE TypeOperators, PolyKinds, DataKinds,
+             TypeFamilies, UndecidableInstances,
              FlexibleInstances, GADTs, FlexibleContexts, ScopedTypeVariables,
-             ConstraintKinds, IncoherentInstances #-}
+             ConstraintKinds, IncoherentInstances, FunctionalDependencies #-}
```
Comment (jmorag) [inline src/Data/Type/Map.hs:8]: `KindSignatures` is implied by `TypeFamilies` and `MultiParamTypeClasses` is implied by `FunctionalDependencies`

**Comment by jmorag** [comment]
Actually, never mind the update thing. I consider it better, but it changes the number of parameters to the `Updatable` class, which is a breaking change that shouldn't be included in this PR. Let me know if you'd like me to open a separate PR with that.

**Comment by jmorag** [comment]
Ping @dorchard

**Comment by jship** [comment]
This change is very useful in some stuff I'm building on top of `type-level-sets`.  Thank you, @jmorag!

**Pull request merged**